xarray/core/accessor_str.py

```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        if dim is None:
+            return self._apply(_get_res_single)
+        else:
+            # dtype MUST be object or strings can be truncated
+            # See: https://github.com/numpy/numpy/issues/8352
+            return self._apply(
+                _get_res_multi,
+                dtype=np.object_,
+                output_core_dims=[[dim]],
+                output_sizes={dim: pat.groups},
+            ).astype(self._obj.dtype.kind)
+
+    def extractall(
+        self,
+        pat: Any,
+        group_dim: Hashable,
+        match_dim: Hashable,
+        case: bool = None,
+        flags: int = 0,
+    ) -> Any:
+        """
+        Extract all matches of capture groups in the regex pat as new
+        dimensions in a DataArray.
+
+        For each string in the DataArray, extract groups from all matches
+        of regular expression pat.
+        Equivalent to applying re.findall() to all the elements in the DataArray
+        and splitting the results across dimensions.
+
+        Parameters
+        ----------
+        pat : str or re.Pattern
+            A string contain a regular expression or
+            a compiled regular expression object.
+        group_dim: hashable
+            Name of the new dimensions corresponding to the capture groups.
+            This dimension is added to the new DataArray first.
+        match_dim: hashable
+            Name of the new dimensions corresponding to the matches for each group.
+            This dimension is added to the new DataArray second.
+        case : bool, default: True
+            If True, case sensitive.
+            Cannoy be set if `pat` is a compiled regular expression.
+        flags : int, default: 0
+            - re module flags, e.g. re.IGNORECASE. Use ``0`` for no flags.
+            - Cannot be set if `pat` is a compiled regex
+
+        Returns
+        -------
+        extracted : same type as values or object array
+
+        Raises
+        ------
+        ValueError : `pat` has no capture groups.
+        ValueError : `case` is set when `pat` is a compiled regular expression.
+        KeyError : Either of the given dimensions is already present in the DataArray.
+        KeyError : The given dimensions names are the same.
+
+        Examples
+        --------
+        Create a string array
+
+        >>> value = xr.DataArray(
+        ...     [
+        ...         [
+        ...             "a_Xy_0",
+        ...             "ab_xY_10-bab_Xy_110-baab_Xy_1100",
+        ...             "abc_Xy_01-cbc_Xy_2210",
+        ...         ],
+        ...         [
+        ...             "abcd_Xy_-dcd_Xy_33210-dccd_Xy_332210",
+        ...             "",
+        ...             "abcdef_Xy_101-fef_Xy_5543210",
+        ...         ],
+        ...     ],
+        ...     dims=["X", "Y"],
+        ... )
+
+        Extract matches
+
+        >>> value.str.extractall(r"(\\w+)_Xy_(\\d*)", group_dim="XX", match_dim="YY")
```

Comment:
```suggestion
        >>> value.str.extractall(r"(\\w+)_Xy_(\\d*)", group_dim="group", match_dim="match")
```
I'd find that easier to understand.

Reply:
I fixed up this issue throughout the examples.